[SC vote is needed] collection_requirements.rst: add a requirement to test against devel or milestone (#3511)
ansible/ansible-documentation: docs/docsite/rst/community/collection_contributors/collection_requirements.rst

* Subscribe to the `news-for-maintainers <https://github.com/ansible-collections/news-for-maintainers>`_ repository.
* Join the `Collection Maintainers & Contributors <https://forum.ansible.com/g/CollectionMaintainer>`_ forum group.

* You MUST run the ``ansible-test sanity`` command against the ``devel`` or ``milestone`` branch of ``ansible-core`` so that you find out about new linting requirements earlier.

> * You MUST also run the ``ansible-test sanity`` command against each of the "major versions" (2.17, 2.18, etc) of ``ansible-core`` that the collection supports. (Usually the ``HEAD`` of the stable-xxx branches.)

Reviewer: ```suggestion
* You MUST also run the ``ansible-test sanity`` command against each of the major feature releases of ``ansible-core`` that the collection supports (2.x, where x denotes a feature release). (Typically, the stable-xxx branches' ``HEAD``.)
```

Author: @Akasurde I like your wording, thanks! Could you please remove "4"? The thing is that they can support more feature releases, EOL too, and they must test against them all

Reviewer: I agree.

Author: Committing, thanks!

  * All entries in ``ignore-*.txt`` files MUST have a justification in a comment in the files for each entry. For example ``plugins/modules/docker_container.py use-argspec-type-path # uses colon-separated paths, can't use type=path``.

> * You MUST run all CI checks against each of the "major versions" (2.17, 2.18, etc) of ``ansible-core`` that the collection supports. (Usually the ``HEAD`` of the stable-xxx branches.)

Reviewer 2: Does this conflict with the statement above?

Here we say that the CI checks must run against `each of the "major versions"` but above we say:

> ... ansible-test sanity command against each of the major feature releases of ``ansible-core`` **that the collection supports**.

What if my collection doesn't support one of the major versions (for whatever reason), as a collection maintainer do I need to test all of them or just the major versions that my collection supports?

Author: 1. I believe it must support all currently supported ansible-core versions and resp. run tests again them all
2. If your collection support EOL versions of core, it must also run tests against them.

I've added some clarifications.

Thanks for asking! I think we should optimize the whole paragraph by re-formatting / re-wording it preserving the meaning in another PR where we can skip polling if the meaning stays the same and if a few SC just approve changes.
Let's now focus on adding the requirement here.

>   - You MUST also run them against the ``devel`` or ``milestone`` branch of ``ansible-core`` at least on a scheduled basis and at least once per week. This ensures that you will find out about new linting requirements and compatibility issues between your collection and ``ansible-core`` earlier.

Reviewer 2: nit

```
... ``ansible-core`` on a scheduled basis of at least once per week.
```

Author: @dbrennand could you suggest the edit please? I'll commit it

Reviewer 2: ```suggestion
  - You MUST also run them against the ``devel`` or ``milestone`` branch of ``ansible-core`` in every PR or on a scheduled basis of at least once per week. This ensures that you will find out about new linting requirements and compatibility issues between your collection and ``ansible-core`` earlier.
```

Author: @dbrennand committed, thanks, I've edited the suggestion, added "in every PR or on s scheduled basis" because it's also an option. FYI

>   - You MUST also run them against the ``devel`` or ``milestone`` branch of ``ansible-core`` in every PR or on a scheduled basis of at least once per week. This ensures that you will find out about new linting requirements and compatibility issues between your collection and ``ansible-core`` earlier.

Reviewer 3: ```suggestion
  - You MUST also run them against the ``devel`` or ``milestone`` branch of ``ansible-core`` in every PR or on a scheduled basis of at least once per week. This ensures that you will find out about new linting requirements and compatibility issues between your collection and ``ansible-core`` early.
```

I think early instead of earlier is better. But I'm not sure, I'm not a native speaker. Maybe someone who is could have a look at this.

Author: SGTM, committed, thanks, PTAL

  * All entries in ``ignore-*.txt`` files MUST have a justification in a comment in the files for each entry. For example ``plugins/modules/docker_container.py use-argspec-type-path # uses colon-separated paths, can't use type=path``.

> * You MUST run CI against each of the "major versions" (2.14, 2.16, 2.17, etc) of ``ansible-core`` that the collection supports. (Usually the ``HEAD`` of the stable-xxx branches.)

Reviewer 4: Please re-add this. Dropping this reduces the requirements. (The similar statement that was added further up is *only* for sanity tests.)

Author: @felixfontein I've put it back, thanks.
~~I think we should remove this statement in future because it can confuse. As we require only sanity tests, to require to run other tests (like units or integration for people who don't know) which are optional feels strange to me, i.e. to force maintainers to run them against all collection-supported core versions just because they have them. If there's an agreement to do it now, i would remove it now not to conduct many votes, but in case of any doubts we can have another dedicated discussion and vote, no problem.~~

UPDATE: i saw the discussion. See https://github.com/ansible/ansible-documentation/pull/3511/changes/7eb095121d116fc5c12d79365734a9c3cdcb6846
